### PR TITLE
feat: introduce `CasePathResolver`

### LINKED ISSUE #22: [FEAT] CasePathResolver: add a method that returns the "full path" i.e compute both Shapes and Edges
**DISCLAIMER**: this issue describes path computation for a single instance/case only. So the implementation will go to the `CasePathResolver` class as it differs from `PathResolver` which is a generic implementation that both deals with single instance or all instances of a process.

### Is your feature request related to a problem? Please describe.
Currently, the existing method of `PathResolver` considers only flowNodes/shapes in the input parameters and returns only the visited flows/edges.

### Describe the solution you'd like
Provide a new method dedicated to the management of a single case instance (create a new `CasePathResolver` class) that returns the "full path" including:
- pass ids of elements that are considered as "completed". For more information, see  #16
- the model elements objects (as `BpmnSemantic`) corresponding to the ids passed as inputs that exist in the model
- considering both flowNodes/shapes and flows/edges as input to compute the path
- computed list of flows/edges and flowNodes/shapes inferred from the inputs

This will require to explicitly split the data returned by the method to identify the input (objects retrieved from the provided ids) from the computed elements. See also #24.

### Additional context
We could use it in the "predictions" demo. We are not using `PathResolver` in the demo for now as it is not considering edge in the input (https://github.com/process-analytics/bpmn-visualization-examples/pull/520). See https://github.com/process-analytics/bpmn-visualization-examples/issues/573.
See also https://github.com/process-analytics/bonita-day-demo-2023/issues/122

### PR DESCRIPTION
**Pull request by tbouffard**
This new class is dedicated to path resolution of a single instance/case of a process.

Given a set of elements considered as completed, it is currently able to compute the edges between the provided shapes and the shapes around the provided edges.
This is the first step towards the implementation of more intelligent computing in the future.

In addition
  - refactor the constructor of `PathResolver`: it now only accepts a `ElementsRegistry` parameter Previously, it required a `BpmnElementsRegistry` object which is not needed as the implementation only needs to access to the methods related to the model.
  - main README: add link to the sources of the demo

Covers #16
Closes #22
Covers #24

### Notes about the types introduced in this PR

The types of the parameter and the returned value are defined to correspond to the state of the final implementation as it is known today. There should be no discussion of types in this PR as the API is not finalized.
What follows is the current proposal with an overview of the structure. It does not mean that the final implementation will be like this (in particular, do not debate now about a pseudo duplication in the type definition, it is too early as the types may change).

```ts
export type CasePathResolverInput = {
    /**
     * The IDs of elements (flowNodes/shapes and flows/edges) that are already completed. Non-existing ids will be silently ignored.
     *
     * `Completed` means that the element has been fully executed or definitively cancelled (for BPM engine that supports it and allow instance continuation of cancelled elements).
     * No further user action or automation will update the element.
     */
    completedIds: string[];

    // see see https://github.com/process-analytics/bv-experimental-add-ons/issues/16
    pendingIds?: string[];
};
```

```ts
export type CasePathResolverOutput = {
    /**
     * The `BpmnSemantic` objects retrieved from the model that relate to the ids passed in {@link CasePathResolverInput}.
     */
    provided: {
        completed: {
            shapes: ShapeBpmnSemantic[];
            edges: EdgeBpmnSemantic[];
        };
        // It is not sure that we will let configure pending edges, see https://github.com/process-analytics/bv-experimental-add-ons/issues/16
        pending?: {
            shapes: ShapeBpmnSemantic[];
            edges: EdgeBpmnSemantic[];
        };
    };
    computed: {
        completed: {
            shapes: ShapeBpmnSemantic[];
            edges: EdgeBpmnSemantic[];
        };
        // This may require a more complex type if we want to provide information about why the elements are considered as candidates only
        // We may also split pending and completed elements
        // See https://github.com/process-analytics/bv-experimental-add-ons/issues/24
        candidates: {
            shapes: ShapeBpmnSemantic[];
            edges: EdgeBpmnSemantic[];
        };
    };
};
```

### CODE REVIEW

File: packages/addons/src/paths.ts [hunk 1]
```diff
@@ -14,30 +14,107 @@ See the License for the specific language governing permissions and
...
+      },
+    };
   }
 }
+
+export type CasePathResolverInput = {
+  /**
+   * The IDs of elements (flowNodes/shapes and flows/edges) that are already completed. Non-existing ids will be silently ignored.
+   *
+   * `Completed` means that the element has been fully executed or definitively cancelled (for BPM engine that supports it and allow instance continuation of cancelled elements).
```
Comment (tbouffard) [inline packages/addons/src/paths.ts:98]:
```suggestion
   * `Completed` means that the element has been fully executed or definitively cancelled (for BPM engines that support this and allow cancelled elements to be continued).
```

File: packages/addons/src/paths.ts [hunk 2]
```diff
@@ -14,30 +14,107 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-import type { BpmnElementsRegistry, ShapeBpmnSemantic } from 'bpmn-visualization';
+import type { EdgeBpmnSemantic, ElementsRegistry, ShapeBpmnSemantic } from 'bpmn-visualization';
+
+// bpmn-visualization does not filter duplicates when passing an ids several times
+const filterDuplicates = (ids: string[]): string[] => [...new Set(ids)];
```
Comment (tbouffard) [inline packages/addons/src/paths.ts:20]: **thought:** IMHO this is a bug in `bpmn-visualization` that we should fix. It applies to all methods of ElementsRegistry taking ids or kinds.
What do you think @csouchet?
If so, I will create an issue.

Comment (csouchet) [reply packages/addons/src/paths.ts:20]: Indeed, the APIs of `bpmn-visualization` should check une unicity of the parameter.

**Review by csouchet — APPROVED** [review]

File: packages/addons/src/paths.ts (hunk 2 above)
Comment (tbouffard) [reply packages/addons/src/paths.ts:20]: Fine, prior to merge this PR, I am going to create a dedicated issue in `bpmn-visualization` and another one in `bv-experimental-add-ons` to update the implementation when the fix is available in a new release of `bpmn-visualization`.

**Comment by tbouffard** [comment]
Putting in draft, waiting for creating issues related to duplicated results returned by the `bpmn-visualization` API.

File: packages/addons/src/paths.ts (hunk 2 above)
Comment (tbouffard) [reply packages/addons/src/paths.ts:20]: Bug created: https://github.com/process-analytics/bpmn-visualization-js/issues/2921
Refactor task created: #150

**Pull request merged**